### ISSUE TITLE
Enhance Buffer to minimise calls to `enableStomping` / `assumeSafeAppend`

### ISSUE THREAD

**Issue by gavin-norman-sociomantic** [sociomantic-tsunami/ocean#85]
Code like the following is very common:
```D
void doSomethingWithMyArray ( ref void[] buf )
{
  buf.length = 0;
  enableStomping(buf);
  fillArrayWithSomething(buf);
}
```
Live testing of calling such code in a tight loop has revealed that the impact of `enableStomping` (which, in turn, calls `assumeSafeAppend`) on the D2 build is significant. (One concrete example: iterating over a large database in memory and writing it out to disk was about 20% slower in the D2 build.)

One way to fix this is to never reduce the length of the referenced array and separately keep track of the actual content. (This is what [`AppendBuffer`](https://github.com/sociomantic-tsunami/ocean/blob/v2.x.x/src/ocean/util/container/AppendBuffer.d) does, by the way.)

`Buffer` already handles a lot of D1/D2 array optimisation concerns (e.g. automatically calling `enableStomping`, where appropriate). It could probably be enhanced to support this optimisation as well, essentially merging `AppendBuffer` into `Buffer`. (We could then deprecate `AppendBuffer`.)

**Comment by mihails-strasuns-sociomantic** [comment]
Note that this requires storing extra manual `capacity` field which makes performance _worse_ in some other cases (afair @mathias-lang-sociomantic checked that when investigating original `ApendBuffer`). In my opinion best approach would be to make it compile-time flag of `Buffer` to chose between two behaviours, though not sure yet which to make default.

**Comment by mathias-lang-sociomantic** [comment]
Maybe we could look into making it a sticky attribute, so we only call it once ?

**Comment by mihails-strasuns-sociomantic** [comment]
Interesting idea, though may required some lobbying in upstream to accept it.

**Comment by nemanja-boric-sociomantic** [comment]
I think @mathias-lang-sociomantic is referring to `ocean.core.Buffer` having a sticky attribute you set once ?

**Comment by mathias-lang-sociomantic** [comment]
> I think @mathias-lang-sociomantic is referring to ocean.core.Buffer having a sticky attribute you set once ?

Yeah, something that would be set by `Buffer`. But it needs to inform the GC about it, which still requires upstream support.

**Comment by mihails-strasuns-sociomantic** [comment]
So, I did some research last week to see what can be done in runtime to support it. My findings so far:

1) Contrary to my previous beliefs, there is not actual field like capacity in metadata (which now makes sense as it differs in slices of same array) but instead a special-case in `_d_arraysetcapacity` to return `0` - https://github.com/dlang/druntime/blob/0de554a84091fb7094d46508c415dab66eb9b86e/src/rt/lifetime.d#L814 Basically whenever it sees that slice length doesn't match initial array length, it considers capacity to be 0.

2) That explains why `enableStomping` call can be so expensive, it ends up with calling https://github.com/dlang/druntime/blob/0de554a84091fb7094d46508c415dab66eb9b86e/src/rt/lifetime.d#L653 (`_d_arrayshrinkfit`) to SHRINK actual allocated array to the length of the slice, not other way around!

3) Most efficient way to implement what we need seems to introduce new attribute flag to `BlkInfo` https://github.com/dlang/druntime/blob/0de554a84091fb7094d46508c415dab66eb9b86e/src/core/memory.d#L227 , for example `ALLOW_STOMPING` and check it to allow appending even if slice length mismatches allocated length.

4) That will require non-trivial refactoring though, because currently there is no single spot which is responsible for checking array capacity - relevant conditions are copy-pasted in several runtime functions that have to take care of it.

**Comment by mihails-strasuns-sociomantic** [comment]
As a totally different approach - what do you think about making `Buffer` strictly malloc based?

**Comment by nemanja-boric-sociomantic** [comment]
It would work only if you either keep the buffers alive or if you're able to figure out their lifetime, right?

**Comment by mihails-strasuns-sociomantic** [comment]
Yeah, but it is already the case as we allow uncontrolled borrowing of buffer content via `opSlice` - GC does not protect you from data being overwritten.

**Comment by nemanja-boric-sociomantic** [comment]
No, but it helps you no to leak memory for the buffers that have no more references.

**Comment by mihails-strasuns-sociomantic** [comment]
Buffer itself is a struct and in D2 structs do have destructors ;)

**Comment by nemanja-boric-sociomantic** [comment]
Right, that's what I was asking myself - if the D2 struct desctructors are enough. We could always version this, though.

**Comment by mihails-strasuns-sociomantic** [comment]
Let's talk about this @ next meeting
